Mainnet/NonFungibleToken/NonFungibleToken/NonFungibleToken.cs

```diff
@@ -0,0 +1,639 @@
+    private void CanTransfer(ulong tokenId)
+    {
+        Address tokenOwner = GetIdToOwner(tokenId);
+        Assert(
+          tokenOwner == this.Message.Sender
+          || GetIdToApproval(tokenId) == Message.Sender
+          || GetOwnerToOperator(tokenOwner, Message.Sender)
+        );
+    }
+
+    /// <summary>
+    /// Guarantees that tokenId is a valid Token.
+    /// </summary>
+    /// <param name="tokenId">ID of the NFT to validate.</param>
+    private void ValidNFToken(ulong tokenId)
+    {
+        Address tokenOwner = GetIdToOwner(tokenId);
+        EnsureAddressIsNotEmpty(tokenOwner);
+    }
+
+    /// <summary>
+    /// Sets the contract owner who can mint/bur
+    /// </summary>
+    /// <param name="owner"></param>
+    public void TransferOwnership(Address owner)
+    {
+        EnsureOwnerOnly();
+        Assert(owner != Address.Zero, $"The {nameof(owner)} parameter can not be default(zero) address.");
+
+        Log(new OwnershipTransferedLog { PreviousOwner = this.Owner, NewOwner = owner });
+
+        this.Owner = owner;
+    }
+
+    private void EnsureOwnerOnly()
+    {
+        Assert(Message.Sender == Owner, "Only owner of the contract can set new owner.");
+    }
+
+    /// <summary>
+    /// Mints new tokens
+    /// </summary>
+    /// <param name="address">The address that will own the minted NFT</param>
+    /// <param name="amount">Number of tokens will be created</param>
+    public void MintAll(Address address, ulong amount)
+    {
+        EnsureOwnerOnly();
+        EnsureAddressIsNotEmpty(address);
+        Assert(amount > 0, "the amount should be higher than zero");
+
+        var index = TotalSupply;
+        var lastIndex = checked(index + amount);
+        var tokenId = NextTokenId;
+
+        while (index < lastIndex)
+        {
+            AddNFToken(address, tokenId);
+            SetTokenByIndex(index, tokenId);
+            SetIndexByToken(tokenId, index);
+
+            Log(new TransferLog { From = Address.Zero, To = address, TokenId = tokenId });
+            
+            checked
+            {
+                index++;
+                tokenId++;
+            }
+        }
+
+        TotalSupply = checked(TotalSupply + amount);
+        NextTokenId = tokenId;
+    }
+
+    public void Burn(ulong tokenId)
+    {
+        Address tokenOwner = GetIdToOwner(tokenId);
+
+        EnsureAddressIsNotEmpty(tokenOwner);
+
+        Assert(tokenOwner == Message.Sender, "Only token owner can burn the token.");
+
+        ClearApproval(tokenId);
+        RemoveNFToken(tokenOwner, tokenId);
+
+        //move last token to removed token and delete last token info
+        var index = GetIndexByToken(tokenId);
+        var lastTokenIndex = checked(--TotalSupply);
+        var lastToken = GetTokenByIndex(lastTokenIndex);
+
+        SetTokenByIndex(index, lastToken);
+        SetIndexByToken(lastToken, index);
+
+        ClearTokenByIndex(lastTokenIndex);
+        ClearIndexByToken(tokenId);
```

Comment:
Should there be a log that a token has been moved to a different index? How would an owner know when their tokens index might have been changed due to a Burn?

Reply:
Only owner can burn his own token already

Reply:
Yes but then the `lastToken` is assigned to the index of the token that was just cleared and burned right? So if I'm the owner the the last token (index 99), and you burn your token (index 50), I get assigned to your old tokens index, how will I know? I'll still be calling the contract looking for index 99 token right?